arangod/RestHandler/RestReplicationHandler.cpp

```diff
@@ -2691,12 +2696,28 @@ void RestReplicationHandler::handleCommandSetTheLeader() {
     }
 
     if (followingTermId.isNumber()) {
-      col->followers()->setTheLeader(
-          leaderId + "_" +
-          StringUtils::itoa(followingTermId.getNumber<uint64_t>()));
+      newLeader =
+          absl::StrCat(leaderId, "_", followingTermId.getNumber<uint64_t>());
     } else {
-      col->followers()->setTheLeader(leaderId);
+      newLeader = leaderId;
     }
+  } else {
+    // no leader change, but maybe the session id for the same leader has
+    // changed
```

Comment:
Do we have to do a `checkPlanLeaderDirect(col, leaderId);` in this branch as well, or is it not necessary? The leader might have changed in the plan already.

Reply:
Probably yes, but I think there is still a potential for a race even with the additional check in place.
Assuming the following sequence of events:
1. new leaderId == current leaderId
2. call to checkPlanLeaderDirect(...), and the leader hasn't changed in the plan
3. leader gets changed
4. we carry out the modification in the FollowerInfo

I think to avoid a race in the new case, we should add a function to FollowerInfo that only changes the session id in case the server id has not changed.

Reply:
I think I am not yet quite sure what the check accomplishes in the original code. Does it prevent a bug, or is it a best-effort check to improve the likelihood of the leadership change succeeding? And do we have to, or want to, accomplish the same here - or is it different?

Reply:
@goedderz : I have added some more guardrails.